projects/pastanaga-angular/src/lib/controls/base-control.ts

```diff
@@ -1,46 +1,433 @@
-import { Input, OnDestroy, OnInit, Directive } from '@angular/core';
+import {
+    ChangeDetectorRef,
+    Directive,
+    Input,
+    OnChanges,
+    OnDestroy,
+    OnInit,
+    Optional,
+    Self,
+    SimpleChanges,
+} from '@angular/core';
+import {
+    FORM_CONTROL,
+    FORM_CONTROL_NAME,
+    IErrorMessages,
+    InternalMode,
+    NG_MODEL,
+    STANDALONE,
+    UpdateOnStrategy,
+} from './form-field.model';
+import {
+    buildAlwaysFalseValidator,
+    concatAllErrorMessages,
+    findFirstErrorMessage,
+    isFormControl,
+    isFormControlName,
+    isNgModel,
+    isStandalone,
+} from './form-field.utils';
+import { merge, Subject } from 'rxjs';
+import {
+    AbstractControl,
+    ControlValueAccessor,
+    FormControl,
+    FormControlDirective,
+    FormControlName,
+    NgControl,
+    ValidatorFn,
+} from '@angular/forms';
 import { coerceBooleanProperty } from '@angular/cdk/coercion';
-import { Subject } from 'rxjs';
+import { markForCheck } from '../common';
+import { takeUntil } from 'rxjs/operators';
 
 let nextId = 0;
 
+/**
+ * Wraps a value, a ngModel or a formControl.
+ * handle id/name generation when not provided.
+ * apply and synchronize internal state to parentComponent NgControl's state
+ */
 @Directive()
-export class BaseControl implements OnInit, OnDestroy {
-    @Input() id?: string;
-    @Input() name?: string;
-    @Input()
-    get help(): string {
-        return this._help;
+export abstract class BaseControl implements OnChanges, OnInit, OnDestroy, ControlValueAccessor {
+    terminator: Subject<void> = new Subject();
+
+    /**
+     * When id of formFieldBase is provided
+     * htmlElement will have a concatenation of this id and fieldKind
+     * otherwise, html element will have a generated id
+     */
+    @Input() set id(value: string | undefined) {
+        this._id = value;
+    }
+
+    /**
+     * When name of formFieldBase is provided
+     * htmlElement should not have name property and should not be considered as part of a form's value
+     * otherwise, html element will have a generated name
```

Comment:
I changed the comment. Basically if we provide a name and we place that same name on the html input element, then we'll have to elements with same name in the dom... which we want to avoid...